python/graphstorm/dataloading/dataloading.py

```diff
@@ -887,3 +893,174 @@ def __iter__(self):
+        # not too large so that workers_per_node * #DataProvider *
+        # num_prefetches * data_shard_size < shm_size.
+
+        self.is_train = is_train
+        assert dataset_path is not None, "dataset_path needs to be specified."
+        if not isinstance(dataset_path, str):
+            raise TypeError(
+                f"dataset_path should be a str, but got {type(dataset_path)} "
+                f"dataset_path={dataset_path}"
+            )
+        file_sampler = DistributedFileSampler(
+            dataset_path=dataset_path,
+            shuffle=shuffle,
+            local_rank=local_rank,
+            world_size=world_size,
+            is_train=is_train,
+        )
+
+        self.file_sampler = file_sampler
+        self.file_sampler_iter = iter(self.file_sampler)
+        self.dataloader_generator = dataloader_generator
+        self.dataloader = None
+        self.data_file = None
+
+        logging.info("DataProvider - Initialization: num_files = %s", len(file_sampler))
+
+    def get_iterator_name(self):
+        """ Return shard name.
+        """
+        return self.data_file
+
+    def refresh_manager(self):
+        """ refresh manager."""
+        self.file_sampler.sampler._index = -1
+
+    def get_iterator(self):
+        """ Get dataloader iterator for a data file.
+        """
+        dataloader = None
+        data_file = next(self.file_sampler_iter)
+
+        # TODO (HZ): implement a queue to store and pop the files by prefetch and get_iterator
+        if data_file is not None:
+            dataloader = self.dataloader_generator.generate(data_file, is_train=self.is_train)
+        self.data_file = data_file
+        self.dataloader = dataloader
+        return dataloader
+
+    def __len__(self):
+        return len(self.file_sampler)
+
+    def __next__(self):
+        return self.get_iterator()
+
+    def __iter__(self):
+        return self
+
+    def release_iterator(self):
+        """ Release the dataloader iterator.
+        """
+        self.dataloader = None
+        self.data_file = None
+
+class DistillDataloaderGenerator:
+    r"""Distill Data Generator that generates pytorch dataloader based on the given file.
+    
+    Parameters:
+    ----------
+    tokenizer : transformers.AutoTokenizer
+        HuggingFace Tokenizer.
+    max_seq_len : int
+        Maximum sequence length.
+    device : str
+        Device name.
+    batch_size : int
+        How many samples per batch to load.
+    collate_fn : func
+        Function to merge a list of samples to form a
+           mini-batch of Tensor(s)
+    """
+
+    def __init__(
+        self,
+        tokenizer,
+        max_seq_len,
+        device,
+        batch_size=1,
+        collate_fn=None,
+    ):
+        self.batch_size = batch_size
+        self.tokenizer = tokenizer
+        self.max_seq_len = max_seq_len
+        self.device = device
+        self.collate_fn = collate_fn
+
+    def _data_len_sync(self, data):
```

Comment:
We need to provide a docstring to explain what does this function do and explain what is the `data`

Reply:
addressed by new commit.